code/headers/gcode_generator.hpp

```diff
@@ -0,0 +1,163 @@
+#pragma once
+#include <hwlib.hpp> // size_t, (u)intx_t
+#include <vector3.hpp>
+
+namespace r2d2::robot_arm {
+    /**
+     * Template Interface class for generating gcodes.
+     * Has a few helper functions such as converting an int to a char.
+     * Size represents the size of the char buffer array.
+     * Note that a string of 4 characters, has a size of 5 with the '\0'
+     * character
+     * @param Size buffer size
+     * */
+    template <size_t Size>
+    class gcode_generator_c {
+    private:
+        /**
+         * Represents the size of the string currently in buffer
+         * */
+        size_t str_len = 0;
+
+        /** Checks whether the string to be added fits into the buffer
+         * @true if it fits
+         * @return false otherwise
+         * */
+        bool string_fits(const char *string) {
+            return get_string_length(string) + 1 + str_len <= Size;
+        }
+
+        /**
+         * Reverses a string
+         *
+         * @param char*
+         * @param int length
+         * */
+        void reverse(char *string, int length) {
+            int start = 0;
+            length--;
```

Comment:
Is length a good name for this parameter? The length does not decrease right? It is only used to get the end point of the string?

Reply:
For the lack of a better variable name, I think we should keep it as length for now since it is still derived from the length of the string.